File: 0402-remove-k-digits/0402-remove-k-digits.java
class Solution {
    public String removeKdigits(String nums, int k) {
        String res = ""; 
        if(nums.length() == k){
            return "0";
        }
        Stack<Character> st = new Stack<>();
        for(int i = 0; i < nums.length(); i++){
            while(!st.isEmpty() && k > 0 && (st.peek() - '0') > (nums.charAt(i)-'0')){
                st.pop();
                k--;
            }
            st.push(nums.charAt(i));
            
        }
        while(k > 0){
            st.pop();
            k--;
        }
        if(st.isEmpty()) return "0";
        while(!st.isEmpty()){
            res += st.pop();
        }
        String ans = "";
        int j;
        for(j = res.length()-1; j > 0; j--){
            if(res.charAt(j) != '0'){
                break;
            }
        }
        for(int i = j; i >=0; i--){
            ans+=res.charAt(i);
        }
        return ans;
    }
}

File: 0402-remove-k-digits/0402-remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        if not num:
            return ""
        if len(num) == k:
            return "0"
        stack = []
        for i in num:
            # stack[-1] = stack.top or stack.peek()
            while(stack and k > 0 and stack[-1] > i):
                stack.pop()
                k -=1
            stack.append(i)
        #if k > 0
        if k > 0:
            stack=stack[:-k]
        else:
            stack
        
        ans=''.join(stack).lstrip('0')
        return ans if ans else "0"